arcconfig/arcconfig.go

```diff
@@ -0,0 +1,164 @@
+	Target string `yaml:"target"`
+}
+
+// AuthenticationRequirement defines the authentication requirement.
+// Precisely one of JWT, Any and All should be set.
+type AuthenticationRequirement struct {
+	// JWT defines the JWTAuthentication
+	JWT JWTAuthentication `yaml:"jwt,omitempty"`
+
+	// Any contains a list of AuthenticationRequirements.
+	// A successful authentication requires one of them being satisfied.
+	Any []AuthenticationRequirement `yaml:"any,omitempty"`
+
+	// All contains a list of AuthenticationRequirements.
+	// A successful authentication requires all of them being satisfied.
+	All []AuthenticationRequirement `yaml:"all,omitempty"`
+}
+
+// JWTAuthentication defines the JWT authentication
+type JWTAuthentication struct {
+	// Name of this JWT Provider, unique within the Proxy.
+	Name string `yaml:"name"`
+
+	// JWT issuer ("iss" claim)
+	Issuer string `yaml:"issuer"`
+
+	// A remote JWKS source
+	RemoteJWKS RemoteJWKS `yaml:"remote_jwks"`
+
+	// Audiences contains a list of audiences
+	Audiences []string `yaml:"audiences,omitempty"`
+
+	// Header name that will contain decoded JWT payload in requests forwarded to
+	// target.
+	ForwardPayloadHeader string `yaml:"forward_payload_header,omitempty"`
+
+	// Locations where JWT may be found. First match wins.
+	In []HTTPParameter `yaml:"in"`
+}
+
+// RemoteJWKS contains information for remote JWKS
+type RemoteJWKS struct {
+	// URL of the JWKS
+	URL string `yaml:"url"`
+
+	// CacheDuration of the JWKS
+	CacheDuration time.Duration `yaml:"cache_duration,omitempty"`
+}
+
+// ConsumerAuthorization is the configuration of API consumer authorization
+type ConsumerAuthorization struct {
+	FailOpen bool `yaml:"fail_open,omitempty"`
+
+	// Locations of API consumer credential (API Key). First match wins.
+	In []HTTPParameter `yaml:"in"`
+
+	// Quota identifier (optional)
+	QuotaIdentifier string `yaml:"quota_identifier,omitempty"`
+}
+
+// HTTPMatch is an HTTP request matching rule
+type HTTPMatch struct {
+	// URL path template using to match incoming requests and optionally identify
+	// path variables. This should be fully denormalized as there is no global
+	// basepath defined anywhere.
+	PathTemplate string `yaml:"path_template"`
+
+	// HTTP method
+	Method HTTPMethod `yaml:"method,omitempty"`
+}
+
+// HTTPParameter defines an HTTP paramter
+// Precisely one of Query, Header or JWTClaim should be set.
+type HTTPParameter struct {
+	// Name of a query paramter
+	Query string `yaml:"query,omitempty"`
+
+	// Name of a header
+	Header string `yaml:"header,omitempty"`
+
+	// A JWTClaim
+	JWTClaim JWTClaim `yaml:"jwt_claim,omitempty"`
+
+	// Prefix to strip off matched value (e.g. "Bearer " for Authorization
+	// tokens).
+	Prefix string `yaml:"prefix,omitempty"`
+}
+
+// JWTClaim has reference to a JWT claim.
+type JWTClaim struct {
+	// Name of the JWT provider
+	Provider string `yaml:"provider"`
+
+	// Name of the claim
+	Name string `yaml:"name"`
+}
+
+type HTTPMethod int
```

Comment:
This will be difficult for human's to read in the config. Perhaps we should use standard http strings instead of arbitrary ints. Doing so would also map to Go's standard http.* method constants.

Reply:
That's true. I forgot that it should also be more human writable. Changed to string and we will validate it upon loading.